File: py_drill/Problems/PQ05_LastStoneWeight.py
# REF : https://leetcode.com/problems/last-stone-weight/
# NOTES :
#       maxHeap application

# T: O(n*logn), logn is of heap
# S: O(n) due to heap

import heapq

class Solution:
    def lastStoneWeight(self, stones: list[int]) -> int:
        maxHeap = [-s for s in stones]
        heapq.heapify(maxHeap)
        while len(maxHeap) > 1:
            first = abs(heapq.heappop(maxHeap))
            second = abs(heapq.heappop(maxHeap))
            if first > second:
                heapq.heappush(maxHeap, -(first - second))

        return abs(maxHeap[0]) if len(maxHeap) == 1 else 0


def main():
    print("output : {}".format(Solution().lastStoneWeight([2,7,4,1,8,1])))

if __name__ == "__main__":
    main()

File: cpp_drill/Problems/PQ05_LastStoneWeight.cpp
#include <iostream>
#include <queue>
using namespace std;

// REF : https://leetcode.com/problems/last-stone-weight/
// NOTES :
//      Max Heap

// T: O(nlogn)
// S: O(n)

class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int> maxHeap;
        for (auto &i : stones)
            maxHeap.push(i);
        
        while (maxHeap.size() > 1) {
            int stoneX = maxHeap.top(); maxHeap.pop();
            int stoneY = maxHeap.top(); maxHeap.pop();
            if (stoneY > stoneX)
                maxHeap.push(stoneY - stoneX);
            if (stoneY < stoneX)
                maxHeap.push(stoneX - stoneY);
        }
        if (maxHeap.size() == 1)
            return maxHeap.top();
        else
            return 0;
    }
};